go/obscuronode/enclave/sql/edbattestation.go

```diff
@@ -0,0 +1,204 @@
+	}
+
+	return certs, report.TCBStatus, verifyErr
+}
+
+func verifyReport(report attestation.Report, cert []byte, config []byte) error {
+	hash := sha256.Sum256(cert)
+	if !bytes.Equal(report.Data[:len(hash)], hash[:]) {
+		return errors.New("report data does not match the certificate's hash")
+	}
+
+	var cfg struct {
+		SecurityVersion uint
+		UniqueID        string
+		SignerID        string
+		ProductID       uint16
+		Debug           bool
+	}
+	if err := json.Unmarshal(config, &cfg); err != nil {
+		return err
+	}
+	if cfg.UniqueID == "" {
+		if cfg.SecurityVersion == 0 {
+			return errors.New("missing securityVersion in config")
+		}
+		if cfg.ProductID == 0 {
+			return errors.New("missing productID in config")
+		}
+	}
+
+	if cfg.SecurityVersion != 0 && report.SecurityVersion < cfg.SecurityVersion {
+		return errors.New("invalid security version")
+	}
+	if cfg.ProductID != 0 && binary.LittleEndian.Uint16(report.ProductID) != cfg.ProductID {
+		return errors.New("invalid product")
+	}
+	if report.Debug && !cfg.Debug {
+		return errors.New("debug enclave not allowed")
+	}
+	if err := verifyID(cfg.UniqueID, report.UniqueID, "unqiueID"); err != nil {
+		return err
+	}
+	if err := verifyID(cfg.SignerID, report.SignerID, "signerID"); err != nil {
+		return err
+	}
+	if cfg.UniqueID == "" && cfg.SignerID == "" {
+		fmt.Println("Warning: Configuration contains neither uniqueID nor signerID!")
+	}
+
+	return nil
+}
+
+func verifyID(expected string, actual []byte, name string) error {
+	if expected == "" {
+		return nil
+	}
+	expectedBytes, err := hex.DecodeString(expected)
+	if err != nil {
+		return err
+	}
+	if !bytes.Equal(expectedBytes, actual) {
+		return errors.New("invalid " + name)
+	}
+	return nil
+}
+
+func httpGetCertQuote(tlsConfig *tls.Config, host, path string) (string, []byte, error) {
+	client := http.Client{Transport: &http.Transport{TLSClientConfig: tlsConfig}}
+	url := url.URL{Scheme: "https", Host: host, Path: path}
+	resp, err := client.Get(url.String()) //nolint:noctx
+	if err != nil {
+		return "", nil, err
+	}
+	body, err := ioutil.ReadAll(resp.Body)
+	var certquote certQuoteResp
+	if err != nil {
+		return "", nil, err
+	}
+
+	/* Newer versions of Marblerun use a common JSON output format in which the quote
+	is embedded into "data" and the error messages are stored inside "message".
+	To keep compatibility with older versions, check if this block exists or
+	if we get the data back directly. */
+
+	if resp.StatusCode != http.StatusOK {
+		errorMessage := gjson.GetBytes(body, "message")
+		if errorMessage.Exists() {
+			return "", nil, errors.New(resp.Status + ": " + errorMessage.String())
+		}
+		return "", nil, errors.New(resp.Status + ": " + string(body))
+	}
+
+	quoteData := gjson.GetBytes(body, "data")
+	if quoteData.Exists() {
+		err = json.Unmarshal([]byte(quoteData.String()), &certquote)
+	} else {
+		err = json.Unmarshal(body, &certquote)
+	}
+
+	if err != nil {
+		return "", nil, err
+	}
+	resp.Body.Close()
```

Comment:
@BedrockSquirrel This should be in a `defer` after the response is created. Also, be sure to check for a `nil` body before attempting to close (it's happened to me before).

Reply:
Yeah good shout.

@joeldudley I don't think you need to check for nil Body if you checked the error before the defer btw. No error should guarantee non-nil resp and resp.Body.

Reply:
@BedrockSquirrel Yeah, but if you check the error before the defer, then you also have to close the resp body when handling the error, so you end up handling the error twice (but you get to skip the nil check, so it balances out).